Add docs for new normalize function. (#784)
neo4j/docs-cypher: modules/ROOT/pages/functions/string.adoc

======

[[functions-normalize]]
== normalize()

> `normalize()` returns the given `STRING` normalized using the `NFC` normalization form.

Reviewer: Added a note to make clear to readers (like me) who are not sure what a normalization form is.

```suggestion
`normalize()` returns the given `STRING` normalized using the `NFC` Unicode normalization form.

[NOTE]
====
Unicode normalization is a process that transforms different representations of the same text into a standardized form.
For more information, see the documentation for link:https://unicode.org/reports/tr15/#Norm_Forms[Unicode normalization forms].
====
```

The `normalize()` function is useful for converting `STRING` values into comparable forms.
When comparing two `STRING` values, it is their Unicode codepoints that are compared.

> In Unicode, a codepoint for a character that looks the same may actually be represented by two, or more, different codepoints.

Reviewer: ```suggestion
In Unicode, a codepoint for a character that looks the same may be represented by two, or more, different codepoints.
```

> For example, the character `<` can be represented as `\uFE64` or `\u003C`. Visually, the character may look the same,
> but if compared, Cypher will return false as `\uFE64` does not equal `\u003C`. Using the `normalize()` function one can
> normalize the codepoint `\uFE64` to `\u003C`, creating a single codepoint representation, allowing them to be successfully compared.

Reviewer: Can we say in brackets which normalization form uses `\uFE64` and `\u003C`?

It feels like something is missing from this sentence: "but if compared, Cypher will return false as `\uFE64` does not equal `\u003C`." Is this because Cypher uses a particular normalization form? and, if so, should we mention what that is?

```suggestion
For example, the character `<` can be represented as `\uFE64` or `\u003C`. Visually, the character may look the same,
but if compared, Cypher will return `false` as `\uFE64` does not equal `\u003C`. Using the `normalize()` function, it is possible to
normalize the codepoint `\uFE64` to `\u003C`, creating a single codepoint representation, allowing them to be successfully compared.
```

Author: It is because \uFE64 != \u003C 😅 not due to Cypher, just because those codes don't match in any way, I guess what this paragraph is trying to explain is that just because a character looks like another, it doesn't mean the encoding is the same. One could think about it like 2 colours, if both are red, they might be different shades or red, our eyes may not pick that up, but the computer knows, as the computer can read the exact "wavelength" so to say, so the computer will say that they are not equal, normalization is a way to make the computer be less fussy and accept that red = red, no matter the shade :)

Reviewer: That makes sense!

*Syntax:*

[source, syntax, role="noheader"]
----
normalize(input)
----

*Returns:*

|===

| `STRING`

|===

*Arguments:*

[options="header"]
|===
| Name | Description

| `input`
| An expression that returns a `STRING`.

|===

*Considerations:*

|===

| `normalize(null)` returns `null`.

|===

.+normalize()+
======

.Query
[source, cypher, indent=0]
----
RETURN normalize('\u212B') = '\u00C5' AS result
----

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===

| +result+
| +true+
1+d|Rows: 1

|===

======

[[functions-normalize-with-normal-form]]
== normalize(), with specified normal form

`normalize()` returns the given `STRING` normalized using the specified normalization form.
The normalization form can be of type `NFC`, `NFD`, `NFKC` or `NFKD`.

> There are two main types of normalization forms. One is based on the concept of canonical equivalence, and the other is based on compatibility.

Reviewer: ```suggestion
There are two main types of normalization forms:
```

> The two forms `NFC` (default) and `NFD` are forms of canonical equivalence. This means that codepoints which represent the same abstract character will
> be normalized to the same codepoint. The same abstract character means that the character has the same visual appearance and behavior.
> The difference between `NFC` and `NFD` is that `NFC` form will always give the *composed* canonical form, one where combined codes are replaced with the single representation, if possible.
> Whereas, `NFD` gives the *decomposed* form, this is the opposite of composed, converting combined codepoints into the split form if possible.

Reviewer: What do we mean by "the combined codes"?

```suggestion
*  *Canonical equivalence*: The `NFC` (default) and `NFD` are forms of canonical equivalence.
This means that codepoints that represent the same abstract character will
be normalized to the same codepoint (and have the same appearance and behavior).
The `NFC` form will always give the *composed* canonical form (in which the combined codes are replaced with a single representation, if possible).
The`NFD` form gives the *decomposed* form (the opposite of the composed form, which converts the combined codepoints into a split form if possible).
```

Author: So a combining unicode is one that will overlay itself on the previous character, e.g for Å this can be just Å or it can be A + a combining char for the little dot. So the codes "combine" together to form one character (visually)

> The two forms `NFKC` and `NFKD` are forms of compatibility normalization. All canonically equivalent sequences are compatible, but not all compatible sequences are canonical.
> This means that a character normalized in `NFC` or `NFD` should also be normalized in `NFKC` and `NFKD`, and other characters that may have a slightly different visual appearance,
> but are considered close enough in appearance to be compatibly equivalent.

Reviewer: Im not sure I understand the last part of this section ("other characters that may have a slightly different visual appearance,
but are considered close enough in appearance to be compatibly equivalent.") 
What other characters are we referring to here?

```suggestion
* *Compatability normalization*: `NFKC` and `NFKD` are forms of compatibility normalization.
All canonically equivalent sequences are compatible, but not all compatible sequences are canonical.
This means that a character normalized in `NFC` or `NFD` should also be normalized in `NFKC` and `NFKD`.
Other characters with only slight differences in appearance should be compatibly equivalent.
```

Author: That is what the example below tries to illustrate, but basically lets say the ﬀ char. This is one Unicode (\uFB00), this char is visually distinguishable from ff (ﬀ != ff : in other fonts this is more obvious, but you may notice if you try to highlight it the first one is one char, and the second is 2 f chars).

Now these have different meanings, two f's are just two letters and can be used to spell a word e.g fluff, but if I spelt fluff as fluﬀ then the length of that word is now 4 instead of 5, because only unicode chars are written. This also is not the two letters ff, same as in Swedish, the Å is not A. So they are not canonically equivalent, as they do not have the same behaviour when used (e.g Å is not pronounced A). But if I am an English speaker, searching a database for your name I would probably spell Åklundh as Aklundh, and would expect the database to still return your name, this is how normalization using compatibility works, whilst I have changed the meaning of your name, in a search way, they are still "compatible" with each other. This type of normalization should not be used to render information, e.g Slack shouldn't normalize your name away, even if the search does :)

======

[[functions-normalize]]
== normalize()

`normalize()` returns the given `STRING` normalized using the `NFC` Unicode normalization form.

[NOTE]
====
Unicode normalization is a process that transforms different representations of the same string into a standardized form.
For more information, see the documentation for link:https://unicode.org/reports/tr15/#Norm_Forms[Unicode normalization forms].
====

The `normalize()` function is useful for converting `STRING` values into comparable forms.
When comparing two `STRING` values, it is their Unicode codepoints that are compared.
In Unicode, a codepoint for a character that looks the same may be represented by two, or more, different codepoints.

> For example, the character `<` can be represented as `\uFE64` (﹤) or `\u003C` (<). Visually, the character may look the same,
> but if compared, Cypher will return false as `\uFE64` does not equal `\u003C`. Using the `normalize()` function, it is possible to
> normalize the codepoint `\uFE64` to `\u003C`, creating a single codepoint representation, allowing them to be successfully compared.

Reviewer: I replaced "Visually" with "To the human eye" to make clearer the distinction you explained in one of your comments. What do you think?

```suggestion
For example, the character `<` can be represented as `\uFE64` (﹤) or `\u003C` (<). 
To the human eye, the characters may appear identical.
However,  if compared, Cypher will return false as `\uFE64` does not equal `\u003C`.
Using the `normalize()` function, it is possible to
normalize the codepoint `\uFE64` to `\u003C`, creating a single codepoint representation, allowing them to be successfully compared.
```

The `NFC` form will always give the *composed* canonical form (in which the combined codes are replaced with a single representation, if possible).
The`NFD` form gives the *decomposed* form (the opposite of the composed form, which converts the combined codepoints into a split form if possible).

* *Compatability normalization*: `NFKC` and `NFKD` are forms of compatibility normalization.
All canonically equivalent sequences are compatible, but not all compatible sequences are canonical.
This means that a character normalized in `NFC` or `NFD` should also be normalized in `NFKC` and `NFKD`.
Other characters with only slight differences in appearance should be compatibly equivalent.

For example, the Greek Upsilon with Acute and Hook Symbol `ϓ` can be represented by the Unicode codepoint: `\u03D3`.

* Normalized in `NFC`: `\u03D3` Greek Upsilon with Acute and Hook Symbol (ϓ)
* Normalized in `NFD`: `\u03D2\u0301` Greek Upsilon with Hook Symbol + Combining Acute Accent (ϓ)
* Normalized in `NFKC`: `\u038E` Greek Capital Letter Upsilon with Tonos (Ύ)
* Normalized in `NFKD`: `\u03A5\u0301` Greek Capital Letter Upsilon + Combining Acute Accent (Ύ)

> In the compatibility normalization forms (`NFKC` and `NFKD`) the character looks visually different as it no longer contains the hook symbol.

Reviewer: ```suggestion
In the compatibility normalization forms (`NFKC` and `NFKD`) the character is visibly different as it no longer contains the hook symbol.
```
